wa/output_processors/postgresql.py

```diff
@@ -0,0 +1,761 @@
+#    Copyright 2018 ARM Limited
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+import psycopg2
+import os
+import uuid
+import re
+import collections
+from psycopg2 import (connect,
+                      extras,
+                      InterfaceError)
+from psycopg2.extensions import AsIs
+
+from devlib.target import KernelVersion, KernelConfig
+from wa import OutputProcessor, Parameter, OutputProcessorError
+from wa.utils.types import level
+from wa.framework.target.info import CpuInfo
+
+
+def cast_level(value, cur):
+    """Generic Level caster for psycopg2"""
+    if value is None:
+        return None
+
+    m = re.match(r"([^\()]*)\((\d*)\)", value)
+    name = str(m.group(1))
+    number = int(m.group(2))
+
+    if m:
+        return level(name, number)
+    else:
+        raise InterfaceError("Bad level representation: {}".format(value))
+
+
+def cast_vanilla(value, cur):
+    """Vanilla Type caster for psycopg2
+
+    Simply returns the string representation.
+    """
+    if value is None:
+        return None
+    else:
+        return str(value)
+
+
+# List functions and classes for adapting
+
+def adapt_level(level):
+    """Generic Level Adapter for psycopg2"""
+    return ("{}({})".format(level.name, level.value))
+
+
+class ListOfLevel(object):
+    value = None
+
+    def __init__(self, level):
+        self.value = level
+
+    def return_original(self):
+        return self.value
+
+
+def adapt_ListOfX(adapt_X):
```

Comment:
All of these need to be redone. "Adapt" is the wrong term here. You're serializing to postrgres format. Break this out into a separate module (or add to `wa/utils/serializer.py` -- though not sure if belongs there) and make sure serialization works propertly -- i.e. according to specification, not just for the inputs you've encountered. Ideally, see if there is already some library you can use for this -- e.g. is there anything in `psychopg2`?

Reply:
I don't think that adapt is the wrong term in this context, since it is the term used in the PostgreSQL documentation (as opposed to Cast, the inverse operation): http://initd.org/psycopg/docs/extensions.html#sql-adaptation-protocol-objects

I will be moving the adapters and the cast functions to a separate module called postgres_convert in utils which I will import in the output processor.